ports/esp32s2/common-hal/busio/I2C.c

```diff
@@ -0,0 +1,223 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/busio/I2C.h"
+#include "py/mperrno.h"
+#include "py/runtime.h"
+
+#include "driver/i2c.h"
+
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "supervisor/shared/translate.h"
+
+typedef enum {
+    STATUS_FREE = 0,
+    STATUS_IN_USE,
+    STATUS_NEVER_RESET
+} i2c_status_t;
+
+static i2c_status_t i2c_status[I2C_NUM_MAX];
+
+void never_reset_i2c(i2c_port_t num) {
+    i2c_status[num] = STATUS_NEVER_RESET;
+}
+
+void i2c_reset(void) {
+    for (i2c_port_t num = 0; num < I2C_NUM_MAX; num++) {
+        if (i2c_status[num] == STATUS_IN_USE) {
+            i2c_driver_delete(num);
+            i2c_status[num] = STATUS_FREE;
+        }
+    }
+}
+
+void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
+        const mcu_pin_obj_t* scl, const mcu_pin_obj_t* sda, uint32_t frequency, uint32_t timeout) {
+    // Pins 45 and 46 are "strapping" pins that impact start up behavior. They usually need to
+    // be pulled-down so pulling them up for I2C is a bad idea. To make this hard, we don't
+    // support I2C on these pins.
+    //
+    // 46 is also input-only so it'll never work.
+    if (scl->number == 45 || scl->number == 46 || sda->number == 45 || sda->number == 46) {
+        mp_raise_ValueError(translate("Invalid pins"));
+    }
+
+#if CIRCUITPY_REQUIRE_I2C_PULLUPS
+    // Test that the pins are in a high state. (Hopefully indicating they are pulled up.)
+    gpio_set_direction(sda->number, GPIO_MODE_DEF_INPUT);
+    gpio_set_direction(scl->number, GPIO_MODE_DEF_INPUT);
+
+    gpio_pulldown_en(sda->number);
+    gpio_pulldown_en(scl->number);
+
+    common_hal_mcu_delay_us(10);
+
+    gpio_pulldown_dis(sda->number);
+    gpio_pulldown_dis(scl->number);
+
+    // We must pull up within 3us to achieve 400khz.
+    common_hal_mcu_delay_us(3);
+
+    if (gpio_get_level(sda->number) == 0 || gpio_get_level(scl->number) == 0) {
+        reset_pin_number(sda->number);
+        reset_pin_number(scl->number);
+        mp_raise_RuntimeError(translate("SDA or SCL needs a pull up"));
+    }
+#endif
+
+
+    self->semaphore_handle = xSemaphoreCreateBinaryStatic(&self->semaphore);
```

Comment:
Do I2C objects live on the GC heap?  If so, and if (when) an I2C object is made long-lived, this internal pointer will point at the non-long-lived version and lead to memory corruption.

Reply:
They do and this would have been a problem. I've removed the handle and just use `&self->semaphore` in its place. (FreeRTOS just passes the pointer back.)